### PR TITLE
Fix test failures introduced by sqlalchemy 1.4.0

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

Fix test failures introduced by sqlalchemy 1.4.0.

## How is this patch tested?

Existing unit tests + Fixed unit tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/db_migrations/versions/39d1c3be5f05_test.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+"""test
```
Comment (harupy) [inline mlflow/store/db_migrations/versions/39d1c3be5f05_test.py:1]: I'm not confidenta if this is the right change 😅

Comment (harupy) [inline mlflow/store/db_migrations/versions/39d1c3be5f05_test.py:1]: Is this the right approach to apply [this change](https://github.com/mlflow/mlflow/pull/4180/files#diff-a7bffc8e5a9a4dd89c47d6f5bd00fd9dbb9f0bd25a82921d5a95f122b4ad3e39R368)?

Comment (harupy) [reply mlflow/store/db_migrations/versions/39d1c3be5f05_test.py:1]: I'll rename this file to a more desriptive name

File: tests/tracking/_tracking_service/test_utils.py [hunk 2]
```diff
@@ -128,7 +128,9 @@ def test_get_store_sqlalchemy_store(tmp_wkdir, db_type):
     env = {_TRACKING_URI_ENV_VAR: uri}
     with mock.patch.dict(os.environ, env), patch_create_engine as mock_create_engine, mock.patch(
         "mlflow.store.db.utils._verify_schema"
-    ), mock.patch("mlflow.store.db.utils._initialize_tables"):
+    ), mock.patch("mlflow.store.db.utils._initialize_tables"), mock.patch(
+        "mlflow.store.tracking.sqlalchemy_store.SqlAlchemyStore.list_experiments", return_value=[]
+    ):
```
Comment (dbczumar) [inline tests/tracking/_tracking_service/test_utils.py:133]: @harupy Can you provide context on why we need to mock `list_experiments()` now, whereas we didn't before?

File: mlflow/store/db_migrations/versions/39d1c3be5f05_test.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/store/db_migrations/versions/39d1c3be5f05_test.py:1]: Looks good! Can we rename the revision (regenerating the revision hash if necessary) to `add_is_nan_constraint_for_latest_metrics_if_necessary` or something similar (feel free to shorten it if that's verbose). 

Can we also insert a comment in the migration explaining that this is only relevant for what I would presume is a small set of users who installed sqlalchemy 1.4.0 with MLflow <= 1.14.1? For all other users (folks with preexisting databases or older sqlalchemy versions), this is a no-op, right?

File: mlflow/store/tracking/dbmodels/models.py [hunk 3]
```diff
@@ -365,7 +365,7 @@ class SqlLatestMetric(Base):
     """
     Step recorded for this metric entry: `BigInteger`.
     """
-    is_nan = Column(Boolean, nullable=False, default=False)
+    is_nan = Column(Boolean(create_constraint=True), nullable=False, default=False)
```
[the remark this answers is not in the document]
Comment (dbczumar) [reply mlflow/store/tracking/dbmodels/models.py:368]: lovely!

File: mlflow/store/tracking/dbmodels/models.py [hunk 4]
```diff
@@ -365,7 +365,7 @@ class SqlLatestMetric(Base):
     """
     Step recorded for this metric entry: `BigInteger`.
```
Comment (dbczumar) [inline mlflow/store/tracking/dbmodels/models.py:366]: @harupy are we also missing `create_constraint=True` for https://github.com/mlflow/mlflow/blob/df2d73e9167a6be93ec903c1be333221d0b2ec81/mlflow/store/tracking/dbmodels/models.py#L311? (`SqlMetric`)

Comment (dbczumar) [reply mlflow/store/tracking/dbmodels/models.py:366]: If so, we should also account for this in your migration above (https://github.com/mlflow/mlflow/pull/4180/files#r596167087).

File: mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py [hunk 5]
```diff
@@ -177,7 +177,7 @@ def upgrade():
             batch_op.drop_constraint(constraint_name="lifecycle_stage", type_="check")
         batch_op.create_check_constraint(
             constraint_name="experiments_lifecycle_stage",
-            condition=column("lifecycle_stage").in_(["active", "deleted"]),
+            condition=column("lifecycle_stage", type_=String).in_(["active", "deleted"]),
```
Comment (dbczumar) [inline mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py:180]: Why do we need to specify `type` now?

Comment (dbczumar) [reply mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py:180]: According to https://github.com/mlflow/mlflow/blob/master/mlflow/temporary_db_migrations_for_pre_1_users/README, these migrations are obsolete. Happy to file a follow-up ticket to remove them.

File: mlflow/store/tracking/dbmodels/models.py (hunk 4 above)
Comment (harupy) [reply mlflow/store/tracking/dbmodels/models.py:366]: https://github.com/mlflow/mlflow/blob/8e74cdd108ff324175e38d621b12d9b2aced05bd/mlflow/store/db_migrations/versions/181f10493468_allow_nulls_for_metric_values.py#L23

Thanks for the catch but I found we set `create_constraint` to `False` for `is_nan` in the `metrics` table.

File: mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py (hunk 5 above)
Comment (harupy) [reply mlflow/temporary_db_migrations_for_pre_1_users/versions/ff01da956556_ensure_unique_constraint_names.py:180]: > Note that this directory will be removed in MLflow 1.1.

Let's remove it :)

File: tests/tracking/_tracking_service/test_utils.py (hunk 2 above)
Comment (harupy) [reply tests/tracking/_tracking_service/test_utils.py:133]: Sure! This is a fix for the error below

https://github.com/mlflow/mlflow/pull/4179/checks?check_run_id=2119720131#step:4:14848

```diff
# error log in pytest
        try:
            (process, labels, extra) = list(
                zip(
                    *[
                        query_entity.row_processor(context, cursor)
-                       for query_entity in context.compile_state._entities
                    ]
                )
            )
- E           ValueError: not enough values to unpack (expected 3, got 0)

# variables defined in this scope
+ compile_state = <MagicMock name='create_engine().connect()._execute_20().context.compiled.compile_state' id='140125509568496'>
context    = <sqlalchemy.orm.context.QueryContext object at 0x7f71828d88b8>
cursor     = <MagicMock name='create_engine().connect()._execute_20()' id='140125495903512'>
filtered   = <MagicMock name='create_engine().connect()._execute_20().context.compiled.compile_state._has_mapper_entities' id='140125462179288'>
single_entity = False
```

It looks like

File: mlflow/store/tracking/dbmodels/models.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/store/tracking/dbmodels/models.py:366]: This is definitely a bit strange! For consistency, can we add an `alter_column` with `create_constraint=True` for the migration you're introducing?

File: tests/tracking/_tracking_service/test_utils.py [hunk 6]
```diff
@@ -128,7 +128,11 @@ def test_get_store_sqlalchemy_store(tmp_wkdir, db_type):
     env = {_TRACKING_URI_ENV_VAR: uri}
     with mock.patch.dict(os.environ, env), patch_create_engine as mock_create_engine, mock.patch(
         "mlflow.store.db.utils._verify_schema"
-    ), mock.patch("mlflow.store.db.utils._initialize_tables"):
+    ), mock.patch("mlflow.store.db.utils._initialize_tables"), mock.patch(
+        # In sqlalchemy 1.4.0, `SqlAlchemyStore.list_experiments` results in an error
```
Comment (dbczumar) [inline tests/tracking/_tracking_service/test_utils.py:132]:
```suggestion
        # In sqlalchemy 1.4.0, `SqlAlchemyStore.list_experiments`, which is called when fetching the store, results in 
        # an error when called with a mocked sqlalchemy engine. Accordingly, we mock `SqlAlchemyStore.list_experiments
```

File: mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_latest_metrics_if_necessary.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
+"""add_is_nan_constraint_for_latest_metrics_if_necessary
```
Comment (dbczumar) [inline mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_latest_metrics_if_necessary.py:1]: Nit: Can we rename this?

```suggestion
"""add_is_nan_constraint_for_metrics_tables_if_necessary
```

Comment (harupy) [reply mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_latest_metrics_if_necessary.py]: Thanks for the catch!

**Comment by harupy** [comment]
We got a new error:

https://github.com/mlflow/mlflow/runs/2162917875#step:4:1724

File: mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_latest_metrics_if_necessary.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+# revision identifiers, used by Alembic.
+revision = "39d1c3be5f05"
+down_revision = "a8c4a736bde6"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # Add a comment
```
Comment (dbczumar) [inline mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_latest_metrics_if_necessary.py:20]: @harupy sounds like a good idea to replace this with an explanatory comment

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks so much @harupy !

I verified that:

- Databases created in sqlalchemy >= 1.14.0 lacking a check constraint can be successfully upgraded using this change on MySQL / sqlite. The upgrade creates check constraints as expected on the `latest_metrics` and `metrics` tables.

- Fresh databases can be created in sqlalchemy 1.13.x / 1.14.x in MySQL and sqlite. The databases contain check constraints as expected.

- Databases created with sqlalchemy 1.13.x can be used after applying the migration introduced in this PR. Check constraints are as expected.

File: tests/store/tracking/test_sqlalchemy_store.py [hunk 9]
```diff
@@ -38,7 +38,7 @@
 from mlflow.utils import mlflow_tags
 from mlflow.utils.file_utils import TempDir
 from mlflow.utils.uri import extract_db_type_from_uri
-from tests.resources.db.initial_models import Base as InitialBase
+from mlflow.store.tracking.dbmodels.initial_models import Base as InitialBase
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store.py:41]: Addressed [this TODO](https://github.com/mlflow/mlflow/pull/4180/files#diff-7e8f3a314f7ba34010ac4d0d42dd1e9714e10c6f7a82158b3c929bb7709521b4L1858)

File: tests/store/tracking/test_sqlalchemy_store_schema.py [hunk 10]
```diff
@@ -21,7 +21,7 @@
     SqlModelVersionTag,
 )
 from mlflow.store.tracking.sqlalchemy_store import SqlAlchemyStore
-from tests.resources.db.initial_models import Base as InitialBase
+from mlflow.store.tracking.dbmodels.initial_models import Base as InitialBase
```
Comment (harupy) [inline tests/store/tracking/test_sqlalchemy_store_schema.py:24]: Addressed [this TODO](https://github.com/mlflow/mlflow/pull/4180/files#diff-7e8f3a314f7ba34010ac4d0d42dd1e9714e10c6f7a82158b3c929bb7709521b4L1858)

File: mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_metrics_tables_if_necessary.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+depends_on = None
+
+
+def upgrade():
+    # This migration is only relevent for users who installed sqlalchemy 1.4.0 with MLflow <= 1.14.1
+    # In sqlalchemy 1.4.0, the default value of `create_costraint` for `sqlalchemy.Boolean` was
+    # changed to False from True:
+    # https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_4_0/lib/sqlalchemy/sql/sqltypes.py#L1841
+    # To preserve the existing check constraints on the `is_nan` columns, explicitly set
+    # `create_constarint` to True.
```
Comment (harupy) [inline mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_metrics_tables_if_necessary.py:25]: @dbczumar Just feel free to fix this comment :)

File: mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_metrics_tables_if_necessary.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    with op.batch_alter_table("latest_metrics") as batch_op:
+        batch_op.alter_column(
+            "is_nan", type_=sa.types.Boolean(create_constraint=True), nullable=False, default=False
```
Comment (dolfinus) [inline mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_metrics_tables_if_necessary.py:28]: MLflow uses alembic<=1.4.1, there is no `default` keyword argument, only `server_default`.

Comment (dolfinus) [reply mlflow/store/db_migrations/versions/39d1c3be5f05_add_is_nan_constraint_for_metrics_tables_if_necessary.py:28]: I've created PR with a fix: https://github.com/mlflow/mlflow/pull/4211

**Pull request closed without merging**